### PR TITLE
GitHub assets should respect render root (fixes #814)

### LINKED ISSUE #814: Github assets should respect the render root
### Expected behavior
With a root of `.` specified, Github assets should be placed in folders relative to that render root

```
lifecycle:
  v1:
    - render:
        root: .
```

### Actual behavior
Github assets are always placed relative to `installer/`.

### Information

- Ship version: master as of b99f3ae5a3c2900d0d53dbb31ac5382f7dec6d0c

### Steps to reproduce the behavior

1. Run ship with the `ship.yaml` included below
2. Observe that Github assets are created relative to `installer/`, not `.`

```
assets:
  v1:
    - github:
        dest: ../base/common
        repo: kubernetes/ingress-nginx
        path: deploy/mandatory.yaml
        ref: nginx-0.22.0
        proxy: false
        strip_path: true

    - github:
        dest: ../base
        repo: kubernetes/ingress-nginx
        path: deploy/provider/aws/service-l4.yaml
        ref: nginx-0.22.0
        proxy: false
        strip_path: true

    - github:
        dest: base
        repo: kubernetes/ingress-nginx
        path: deploy/provider/aws/
        ref: nginx-0.22.0
        proxy: false
        strip_path: true

    - github:
        dest: ../overlays/cloud
        repo: kubernetes/ingress-nginx
        path: deploy/provider/aws/patch-configmap-l4.yaml
        ref: nginx-0.22.0
        proxy: false
        strip_path: true
config:
  v1: []

lifecycle:
  v1:
    - render:
        root: .
    - message:
        contents: "outro"
```

### PR DESCRIPTION
**Pull request by dexhorthy**
What I Did
------------

Ensure github assets respect the render root (fixes #814)

Also, fix a null `planner.Local` when using `withStatusReceiver` [here](https://github.com/replicatedhq/ship/pull/827/files#diff-b9011e89fa3074e9c3fc633d2dc9cd6bR70)

How I Did it
------------

Pass down `renderRoot` from `CLIPlanner#build` down into the `github/render.go`, use it instead of `constants.InstallerPrefixPath`

How to verify it
------------

See example in #814

Description for the Changelog
------------

Ensure github assets respect the render root (fixes #814)

### CODE REVIEW

File: pkg/lifecycle/render/github/render.go [hunk 1]
```diff
@@ -279,7 +281,7 @@ func getDestPathNoProxy(asset api.GitHubAsset, builder *templates.Builder) (stri
 		}
 	}
 
-	return filepath.Join(constants.InstallerPrefixPath, destDir, assetPath), nil
+	return filepath.Join(renderRoot, destDir, assetPath), nil
```
Comment (dexhorthy) [inline pkg/lifecycle/render/github/render.go]: This is the change that fixes it, the rest is passing this down through the call stack. Aware that this value is very much tramp data at the moment, but don't have a solution to that off the top of my head.

**Comment by laverya** [comment]
@dexhorthy do you want to fix the integration tests before I review this?

**Comment by dexhorthy** [comment]
@laverya okay, tests passing. Will review Codeclimate errors but I'm assuming its mostly small stuff

File: pkg/lifecycle/render/render.go [hunk 2]
```diff
@@ -23,8 +23,9 @@ var (
 	ProgressRender = daemontypes.StringProgress("render", "rendering assets and configuration values")
 )
 
-// A renderer takes a resolved spec, collects config values, and renders assets
-type renderer struct {
+// A headlessrenderer takes a resolved spec, collects config values, and renders assets
+// this was the old plain "renderer", but now its only used in headless workflows, so renaming it for clarity
+type headlessrenderer struct {
```
Comment (dexhorthy) [inline pkg/lifecycle/render/render.go]: Renamed this, it's a small step toward removing pre-navcycle bits

File: integration/init/github-no-proxy-gogetter/metadata.yaml [hunk 3]
```diff
@@ -1,3 +1,4 @@
 upstream: "input/ship.yaml"
 make_absolute: true
-args: ["--prefer-git"]
\ No newline at end of file
+args: ["--prefer-git"]
+skip_cleanup: true
```
Comment (laverya) [inline integration/init/github-no-proxy-gogetter/metadata.yaml]: probably not something to leave here

File: pkg/lifecycle/render/render.go [hunk 4]
```diff
@@ -58,16 +59,23 @@ func (r *renderer) Execute(ctx context.Context, release *api.Release, step *api.
...
-	debug.Log("event", "backup.start")
-	err = r.backupIfPresent(constants.InstallerPrefixPath)
-	if err != nil {
-		return errors.Wrapf(err, "backup existing install directory %s", constants.InstallerPrefixPath)
+	if step.Root != "." && step.Root != "./" {
+		debug.Log("event", "backup.start")
+		err = r.backupIfPresent(step.Root)
+		if err != nil {
+			return errors.Wrapf(err, "backup existing install directory %s", constants.InstallerPrefixPath)
+		}
```
Comment (laverya) [inline pkg/lifecycle/render/render.go]: Just a thought - should we check for, and disallow, the use of `..` within the render root?

File: integration/init/github-no-proxy-gogetter/metadata.yaml (hunk 3 above)
Comment (dexhorthy) [reply integration/init/github-no-proxy-gogetter/metadata.yaml]: thanks

File: pkg/lifecycle/render/render.go (hunk 4 above)
Comment (dexhorthy) [reply pkg/lifecycle/render/render.go]: argh yeah. We probably need a proper chroot here someday.

**Pull request merged**